### PR TITLE
scollector: vSphere collector with performance counters

### PR DESCRIPTION
**Pull request by simnv**
New metrics:
- vsphere.perf.\* for host stats
- vsphere.guest.perf.\* for virtual machine stats

Metrics are populated dynamically, names, units, rates and descriptions are pulled from vSphere itself.

Performance metrics are opt-in, you should set include mask in PerformanceMetrics parameter. For example:
  `PerformanceMetrics = ['\.cpu\.ready$', '^vsphere(\.guest)?\.perf\.(cpu|net|power)\.', '^vsphere(\.guest)?\.perf\.mem\.(de)?compression']`

Use PerformanceMetrics = ['.*'] to collect all additional metrics.

Host metrics:

> vsphere.perf.cpu.coreUtilization
> vsphere.perf.cpu.costop
> vsphere.perf.cpu.demand
> vsphere.perf.cpu.idle
> vsphere.perf.cpu.latency
> vsphere.perf.cpu.ready
> vsphere.perf.cpu.reservedCapacity
> vsphere.perf.cpu.swapwait
> vsphere.perf.cpu.totalCapacity
> vsphere.perf.cpu.usage
> vsphere.perf.cpu.usagemhz
> vsphere.perf.cpu.used
> vsphere.perf.cpu.utilization
> vsphere.perf.cpu.wait
> vsphere.perf.datastore.datastoreIops
> vsphere.perf.datastore.datastoreMaxQueueDepth
> vsphere.perf.datastore.datastoreNormalReadLatency
> vsphere.perf.datastore.datastoreNormalWriteLatency
> vsphere.perf.datastore.datastoreReadBytes
> vsphere.perf.datastore.datastoreReadIops
> vsphere.perf.datastore.datastoreReadLoadMetric
> vsphere.perf.datastore.datastoreReadOIO
> vsphere.perf.datastore.datastoreVMObservedLatency
> vsphere.perf.datastore.datastoreWriteBytes
> vsphere.perf.datastore.datastoreWriteIops
> vsphere.perf.datastore.datastoreWriteLoadMetric
> vsphere.perf.datastore.datastoreWriteOIO
> vsphere.perf.datastore.maxTotalLatency
> vsphere.perf.datastore.numberReadAveraged
> vsphere.perf.datastore.numberWriteAveraged
> vsphere.perf.datastore.read
> vsphere.perf.datastore.siocActiveTimePercentage
> vsphere.perf.datastore.sizeNormalizedDatastoreLatency
> vsphere.perf.datastore.totalReadLatency
> vsphere.perf.datastore.totalWriteLatency
> vsphere.perf.datastore.write
> vsphere.perf.disk.busResets
> vsphere.perf.disk.commands
> vsphere.perf.disk.commandsAborted
> vsphere.perf.disk.commandsAveraged
> vsphere.perf.disk.deviceLatency
> vsphere.perf.disk.deviceReadLatency
> vsphere.perf.disk.deviceWriteLatency
> vsphere.perf.disk.kernelLatency
> vsphere.perf.disk.kernelReadLatency
> vsphere.perf.disk.kernelWriteLatency
> vsphere.perf.disk.maxQueueDepth
> vsphere.perf.disk.maxTotalLatency
> vsphere.perf.disk.numberRead
> vsphere.perf.disk.numberReadAveraged
> vsphere.perf.disk.numberWrite
> vsphere.perf.disk.numberWriteAveraged
> vsphere.perf.disk.queueLatency
> vsphere.perf.disk.queueReadLatency
> vsphere.perf.disk.queueWriteLatency
> vsphere.perf.disk.read
> vsphere.perf.disk.totalLatency
> vsphere.perf.disk.totalReadLatency
> vsphere.perf.disk.totalWriteLatency
> vsphere.perf.disk.usage
> vsphere.perf.disk.write
> vsphere.perf.hbr.hbrNetRx
> vsphere.perf.hbr.hbrNetTx
> vsphere.perf.hbr.hbrNumVms
> vsphere.perf.mem.active
> vsphere.perf.mem.activewrite
> vsphere.perf.mem.compressed
> vsphere.perf.mem.compressionRate
> vsphere.perf.mem.consumed
> vsphere.perf.mem.decompressionRate
> vsphere.perf.mem.granted
> vsphere.perf.mem.heap
> vsphere.perf.mem.heapfree
> vsphere.perf.mem.latency
> vsphere.perf.mem.llSwapIn
> vsphere.perf.mem.llSwapInRate
> vsphere.perf.mem.llSwapOut
> vsphere.perf.mem.llSwapOutRate
> vsphere.perf.mem.llSwapUsed
> vsphere.perf.mem.lowfreethreshold
> vsphere.perf.mem.overhead
> vsphere.perf.mem.reservedCapacity
> vsphere.perf.mem.shared
> vsphere.perf.mem.sharedcommon
> vsphere.perf.mem.state
> vsphere.perf.mem.swapin
> vsphere.perf.mem.swapinRate
> vsphere.perf.mem.swapout
> vsphere.perf.mem.swapoutRate
> vsphere.perf.mem.swapused
> vsphere.perf.mem.sysUsage
> vsphere.perf.mem.totalCapacity
> vsphere.perf.mem.unreserved
> vsphere.perf.mem.usage
> vsphere.perf.mem.vmmemctl
> vsphere.perf.mem.zero
> vsphere.perf.net.broadcastRx
> vsphere.perf.net.broadcastTx
> vsphere.perf.net.bytesRx
> vsphere.perf.net.bytesTx
> vsphere.perf.net.droppedRx
> vsphere.perf.net.droppedTx
> vsphere.perf.net.errorsRx
> vsphere.perf.net.errorsTx
> vsphere.perf.net.multicastRx
> vsphere.perf.net.multicastTx
> vsphere.perf.net.packetsRx
> vsphere.perf.net.packetsTx
> vsphere.perf.net.received
> vsphere.perf.net.transmitted
> vsphere.perf.net.unknownProtos
> vsphere.perf.net.usage
> vsphere.perf.power.energy
> vsphere.perf.power.power
> vsphere.perf.power.powerCap
> vsphere.perf.rescpu.actav1
> vsphere.perf.rescpu.actav15
> vsphere.perf.rescpu.actav5
> vsphere.perf.rescpu.actpk1
> vsphere.perf.rescpu.actpk15
> vsphere.perf.rescpu.actpk5
> vsphere.perf.rescpu.maxLimited1
> vsphere.perf.rescpu.maxLimited15
> vsphere.perf.rescpu.maxLimited5
> vsphere.perf.rescpu.runav1
> vsphere.perf.rescpu.runav15
> vsphere.perf.rescpu.runav5
> vsphere.perf.rescpu.runpk1
> vsphere.perf.rescpu.runpk15
> vsphere.perf.rescpu.runpk5
> vsphere.perf.rescpu.sampleCount
> vsphere.perf.rescpu.samplePeriod
> vsphere.perf.storageAdapter.commandsAveraged
> vsphere.perf.storageAdapter.maxTotalLatency
> vsphere.perf.storageAdapter.numberReadAveraged
> vsphere.perf.storageAdapter.numberWriteAveraged
> vsphere.perf.storageAdapter.read
> vsphere.perf.storageAdapter.totalReadLatency
> vsphere.perf.storageAdapter.totalWriteLatency
> vsphere.perf.storageAdapter.write
> vsphere.perf.storagePath.commandsAveraged
> vsphere.perf.storagePath.maxTotalLatency
> vsphere.perf.storagePath.numberReadAveraged
> vsphere.perf.storagePath.numberWriteAveraged
> vsphere.perf.storagePath.read
> vsphere.perf.storagePath.totalReadLatency
> vsphere.perf.storagePath.totalWriteLatency
> vsphere.perf.storagePath.write
> vsphere.perf.sys.resourceCpuAct1
> vsphere.perf.sys.resourceCpuAct5
> vsphere.perf.sys.resourceCpuAllocMax
> vsphere.perf.sys.resourceCpuAllocMin
> vsphere.perf.sys.resourceCpuAllocShares
> vsphere.perf.sys.resourceCpuMaxLimited1
> vsphere.perf.sys.resourceCpuMaxLimited5
> vsphere.perf.sys.resourceCpuRun1
> vsphere.perf.sys.resourceCpuRun5
> vsphere.perf.sys.resourceCpuUsage
> vsphere.perf.sys.resourceFdUsage
> vsphere.perf.sys.resourceMemAllocMax
> vsphere.perf.sys.resourceMemAllocMin
> vsphere.perf.sys.resourceMemAllocShares
> vsphere.perf.sys.resourceMemConsumed
> vsphere.perf.sys.resourceMemCow
> vsphere.perf.sys.resourceMemMapped
> vsphere.perf.sys.resourceMemOverhead
> vsphere.perf.sys.resourceMemShared
> vsphere.perf.sys.resourceMemSwapped
> vsphere.perf.sys.resourceMemTouched
> vsphere.perf.sys.resourceMemZero
> vsphere.perf.sys.uptime
> vsphere.perf.vflashModule.numActiveVMDKs

Virtual machine metrics:

> vsphere.guest.perf.cpu.costop
> vsphere.guest.perf.cpu.demand
> vsphere.guest.perf.cpu.demandEntitlementRatio
> vsphere.guest.perf.cpu.entitlement
> vsphere.guest.perf.cpu.idle
> vsphere.guest.perf.cpu.latency
> vsphere.guest.perf.cpu.maxlimited
> vsphere.guest.perf.cpu.overlap
> vsphere.guest.perf.cpu.ready
> vsphere.guest.perf.cpu.run
> vsphere.guest.perf.cpu.swapwait
> vsphere.guest.perf.cpu.system
> vsphere.guest.perf.cpu.usage
> vsphere.guest.perf.cpu.usagemhz
> vsphere.guest.perf.cpu.used
> vsphere.guest.perf.cpu.wait
> vsphere.guest.perf.datastore.maxTotalLatency
> vsphere.guest.perf.datastore.numberReadAveraged
> vsphere.guest.perf.datastore.numberWriteAveraged
> vsphere.guest.perf.datastore.read
> vsphere.guest.perf.datastore.totalReadLatency
> vsphere.guest.perf.datastore.totalWriteLatency
> vsphere.guest.perf.datastore.write
> vsphere.guest.perf.disk.busResets
> vsphere.guest.perf.disk.commands
> vsphere.guest.perf.disk.commandsAborted
> vsphere.guest.perf.disk.commandsAveraged
> vsphere.guest.perf.disk.maxTotalLatency
> vsphere.guest.perf.disk.numberRead
> vsphere.guest.perf.disk.numberReadAveraged
> vsphere.guest.perf.disk.numberWrite
> vsphere.guest.perf.disk.numberWriteAveraged
> vsphere.guest.perf.disk.read
> vsphere.guest.perf.disk.usage
> vsphere.guest.perf.disk.write
> vsphere.guest.perf.mem.active
> vsphere.guest.perf.mem.activewrite
> vsphere.guest.perf.mem.compressed
> vsphere.guest.perf.mem.compressionRate
> vsphere.guest.perf.mem.consumed
> vsphere.guest.perf.mem.decompressionRate
> vsphere.guest.perf.mem.entitlement
> vsphere.guest.perf.mem.granted
> vsphere.guest.perf.mem.latency
> vsphere.guest.perf.mem.llSwapInRate
> vsphere.guest.perf.mem.llSwapOutRate
> vsphere.guest.perf.mem.llSwapUsed
> vsphere.guest.perf.mem.overhead
> vsphere.guest.perf.mem.overheadMax
> vsphere.guest.perf.mem.overheadTouched
> vsphere.guest.perf.mem.shared
> vsphere.guest.perf.mem.swapin
> vsphere.guest.perf.mem.swapinRate
> vsphere.guest.perf.mem.swapout
> vsphere.guest.perf.mem.swapoutRate
> vsphere.guest.perf.mem.swapped
> vsphere.guest.perf.mem.swaptarget
> vsphere.guest.perf.mem.usage
> vsphere.guest.perf.mem.vmmemctl
> vsphere.guest.perf.mem.vmmemctltarget
> vsphere.guest.perf.mem.zero
> vsphere.guest.perf.mem.zipSaved
> vsphere.guest.perf.mem.zipped
> vsphere.guest.perf.net.broadcastRx
> vsphere.guest.perf.net.broadcastTx
> vsphere.guest.perf.net.bytesRx
> vsphere.guest.perf.net.bytesTx
> vsphere.guest.perf.net.droppedRx
> vsphere.guest.perf.net.droppedTx
> vsphere.guest.perf.net.multicastRx
> vsphere.guest.perf.net.multicastTx
> vsphere.guest.perf.net.packetsRx
> vsphere.guest.perf.net.packetsTx
> vsphere.guest.perf.net.received
> vsphere.guest.perf.net.transmitted
> vsphere.guest.perf.net.usage
> vsphere.guest.perf.power.energy
> vsphere.guest.perf.power.power
> vsphere.guest.perf.rescpu.actav1
> vsphere.guest.perf.rescpu.actav15
> vsphere.guest.perf.rescpu.actav5
> vsphere.guest.perf.rescpu.actpk1
> vsphere.guest.perf.rescpu.actpk15
> vsphere.guest.perf.rescpu.actpk5
> vsphere.guest.perf.rescpu.maxLimited1
> vsphere.guest.perf.rescpu.maxLimited15
> vsphere.guest.perf.rescpu.maxLimited5
> vsphere.guest.perf.rescpu.runav1
> vsphere.guest.perf.rescpu.runav15
> vsphere.guest.perf.rescpu.runav5
> vsphere.guest.perf.rescpu.runpk1
> vsphere.guest.perf.rescpu.runpk15
> vsphere.guest.perf.rescpu.runpk5
> vsphere.guest.perf.rescpu.sampleCount
> vsphere.guest.perf.rescpu.samplePeriod
> vsphere.guest.perf.sys.heartbeat
> vsphere.guest.perf.sys.osUptime
> vsphere.guest.perf.sys.uptime
> vsphere.guest.perf.virtualDisk.largeSeeks
> vsphere.guest.perf.virtualDisk.mediumSeeks
> vsphere.guest.perf.virtualDisk.numberReadAveraged
> vsphere.guest.perf.virtualDisk.numberWriteAveraged
> vsphere.guest.perf.virtualDisk.read
> vsphere.guest.perf.virtualDisk.readIOSize
> vsphere.guest.perf.virtualDisk.readLatencyUS
> vsphere.guest.perf.virtualDisk.readLoadMetric
> vsphere.guest.perf.virtualDisk.readOIO
> vsphere.guest.perf.virtualDisk.smallSeeks
> vsphere.guest.perf.virtualDisk.totalReadLatency
> vsphere.guest.perf.virtualDisk.totalWriteLatency
> vsphere.guest.perf.virtualDisk.write
> vsphere.guest.perf.virtualDisk.writeIOSize
> vsphere.guest.perf.virtualDisk.writeLatencyUS
> vsphere.guest.perf.virtualDisk.writeLoadMetric
> vsphere.guest.perf.virtualDisk.writeOIO

### CODE REVIEW

File: cmd/scollector/collectors/redis_linux.go [hunk 1]
```diff
@@ -1,5 +1,3 @@
-// +build darwin linux
```
Comment (captncraig) [inline cmd/scollector/collectors/redis_linux.go]: Is this a related change? Are you monitoring redis on windows?

Comment (captncraig) [inline cmd/scollector/collectors/redis_linux.go]: Because part of this collector is executing `ps`, which will likely fail on windows.

Comment (simnv) [inline cmd/scollector/collectors/redis_linux.go]: Oh, collector build was broken at the time i was making this PR. Had to patch manually.

File: cmd/scollector/collectors/vsphere.go [hunk 2]
```diff
@@ -61,6 +66,82 @@ type DatastoreHostMount struct {
...
+
+	res, err = v.Datastores(storeKey)
+	if err != nil {
+		return err
+	}
+	for _, r := range res {
+		var name string
+		for _, p := range r.Props {
+			if p.Name == "name" {
+				name, _ = opentsdb.Clean(p.Val.Inner)
```
Comment (captncraig) [inline cmd/scollector/collectors/vsphere.go]: don't ignore errors.

File: cmd/scollector/collectors/vsphere.go [hunk 3]
```diff
@@ -61,6 +66,82 @@ type DatastoreHostMount struct {
...
+		"name",
+	})
+	if err != nil {
+		return err
+	}
+	for _, r := range res {
+		var name string
+		for _, p := range r.Props {
+			if p.Name == "name" {
+				name, _ = opentsdb.Clean(p.Val.Inner)
```
Comment (captncraig) [inline cmd/scollector/collectors/vsphere.go]: don't ignore errors.

File: cmd/scollector/collectors/vsphere.go [hunk 4]
```diff
@@ -61,6 +66,82 @@ type DatastoreHostMount struct {
...
+				break
+			}
+		}
+		if name == "" {
+			continue
+		}
+		storeKey[r.ID] = name
+		for _, p := range r.Props {
+			switch p.Val.Type {
```
Comment (captncraig) [inline cmd/scollector/collectors/vsphere.go]: feels a bit odd to use a switch with only a single case. Perhaps an if will suffice?

File: cmd/scollector/collectors/vsphere.go [hunk 5]
```diff
@@ -61,6 +66,82 @@ type DatastoreHostMount struct {
...
+			case "VmfsDatastoreInfo":
+				var vdistring = fmt.Sprintf("<val xsi:type=\"VmfsDatastoreInfo\">%s</val>", p.Val.Inner)
+				var vdi VmfsDatastoreInfo
+				d := xml.NewDecoder(bytes.NewBufferString(vdistring))
```
Comment (captncraig) [inline cmd/scollector/collectors/vsphere.go]: [`strings.NewReader`](https://golang.org/pkg/strings/#NewReader) would be more appropriate here than the bytes version.

File: cmd/scollector/collectors/vsphere.go (hunk 4 above)
Comment (simnv) [inline cmd/scollector/collectors/vsphere.go]: Had this added for future HostNasVolume and HostLocalFileSystemVolume support, but we don't use those and also there is nothing interesting there, so just deleted it. Will change it, thanks.

**Comment by kylebrandt** [comment]
Did this ever get change to have the extended metrics be off by default?

**Comment by simnv** [comment]
Sorry for the delay, will look into that next week. Had to sort out my priorities.

**Comment by kylebrandt** [comment]
Well, considering my response rate on PRs, you are not alone ;)

**Comment by simnv** [comment]
@kylebrandt Done. Made performance metrics as opt-in.

**Comment by kylebrandt** [comment]
@simnv Sorry this took so long, was pushing to get 0.5.0 out. From my test it looks like this changes the default behavior and no metrics are sent if there is no PerformanceMetrics option. Can this be made so the original metrics don't change any only the new ones are opt-in so there isn't a breaking change?

**Comment by simnv** [comment]
@kylebrandt Don't understand your question. Default behaviour is not changed. Function vsphereSkipPerformanceMetric is called only for performance metrics, it doesn't affect original metrics.

Just tried it myself: if PerformanceMetrics parameter is commented out, original vsphere metrics are sent as usual and with usual pause before sendind - first vsphere connect is slow.

Added small clarification in docs:
> If PerformanceMetrics parameter is not set, you will still get a basic subset of vSphere metrics.

**Comment by kylebrandt** [comment]
@simnv Confirmed, my error, sorry

**Comment by kylebrandt** [comment]
@simnv Confused about the vcenter tag, why set that to the host that scollector is running on since this is polling collector?

**Comment by simnv** [comment]
@kylebrandt Oh, sorry, my mistake to assume that scollector host is the same host with vcenter. Fixed.

We have more than one vcenter and need to group information by its name. The host tag by default is overwritten with esx host name and I didn't want to break anything. Hence the new tag.
